gluten-core/src/main/scala/org/apache/spark/util/GlutenTaskContext.scala

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.util
+
+import org.apache.spark.internal.Logging
+
+import _root_.io.glutenproject.memory.TaskMemoryMetrics
+
+import java.util
+import java.util.Collections
+
+import scala.collection.JavaConverters._
+
+object GlutenTaskContext {
+
+  def get(): TaskResourceRegistry = registry.get
+
+  private[this] val registry: ThreadLocal[TaskResourceRegistry] =
+    new ThreadLocal[TaskResourceRegistry]
+
+  protected[spark] def setTaskContext(tc: TaskResourceRegistry): Unit = registry.set(tc)
+
+  protected[spark] def unset(): Unit = registry.remove()
+}
```

Comment:
What's the pros of adding this thread local storage? Seems to me that we should manage an extra task lifecycle state in Gluten's code. Previously `TaskContext.get()` is exposed in a stateless way (although it manages states internally)

Reply:
> What's the pros of adding this thread local storage? 

Avoid synchronized lock and thread check between across JNI.

> we should manage an extra task lifecycle state in Gluten's code

For what kinds of resources?

Reply:
> For what kinds of resources?

"we should manage an extra task lifecycle state in Gluten's code" -> "within this change, we should manage an extra task lifecycle state in Gluten's code"

Reply:
Is one of the main purpose to remove the "FALLBACK_REGISTRY"?

Reply:
> Is one of the main purpose to remove the "FALLBACK_REGISTRY"?

Yes, Gluten use single thread execution for now, thus all resources/states could be ''thread local''.

Reply:
> "within this change, we should manage an extra task lifecycle state in Gluten's code"

TaskResourceRegistry is bound with TaskContext, so they have same lifecycle.

Reply:
> TaskResourceRegistry is bound with TaskContext, so they have same lifecycle.

OK. FWIW let's take it into consideration that Spark uses this task-thread model:

```
Thread 1 --> TaskContext 1
   |__ (Child) Thread  2   --> TaskContext 1

Thread 3 --> TaskContext 2

Thread 4 --> TaskContext 3

Driver Thread 5 --> No TaskContext (null)
```

To me here exists some tricky parts to optimize this part of code:

1. What happens if we are on a child thread of a task thread? e.g. Thread 2 
2. What happens if we are on the driver thread? e.g. Thread 5
3. What happens if we are not on any thread that are created by Spark, for example, a UT thread?

You can just refer to these questions when doing the optimization, I can get back to the PR when it's ready.

Reply:
Thanks for great examples!